sycl/doc/design/DeviceAspectTraitDesign.md

```diff
@@ -0,0 +1,128 @@
+[DeviceIf.md][4]. Additionally, it leverages part of the design for optional
+kernel features, as described in [OptionalDeviceFeatures.md][5].
+
+## Changes to the compiler driver
+
+Using the `-fsycl-targets` options introduced in [DeviceIf.md][4] and the
+configuration file introduced in [OptionalDeviceFeatures.md][5], the compiler
+driver finds the set of all aspects supported by each specified target. Note
+that in this section we refer to aspects as their integral representation as
+specified in the device headers rather than by the names specified in the
+[SYCL 2020 specification][1].
+
+For each target $t$ in `-fsycl-targets`, let $A^{any}_t$ be the set of aspects
+supported by any device supporting $t$ and let $A^{all}_t$ be the set of aspects
+supported by all devices supporting $t$. These sets are defined as follows:
+* If $t$ has an entry in the configuration file, $A^{all}_t$ is the same as the
+`aspects` list in that entry.
+* If $t$ does not have an entry in the configuration file, $A^{all}_t$ is empty.
+* If $t$ has an entry in the configuration file and the entry has a value
+`may_support_other_aspects` set to `false`, $A^{any}_t$ is the same as the
+`aspects` list in that entry.
+* If $t$ does not have an entry the configuration file or the entry has a value
+`may_support_other_aspects` set to `true`, $A^{any}_t$ is the set of all
+aspects.
+
+For example, the target `intel_gpu_dg1` is supported by a specific device (DG1)
+and as such would have an entry in the configuration file with `aspects` being
+the set of aspects that device supports. Likewise it would have
+`may_support_other_aspects` set to `false` as there will be no other devices
+supporting this target, meaning there will never be any devices supporting
+the target and supporting anything not in `aspects`.  In contrast, the target
+`nvidia_gpu_sm_80` is supported by CUDA devices with `sm_80` architecture or
+newer, so its entry in the configuration file would have
+`may_support_other_aspects` set to `true` to indicate that there could be future
+devices that support aspects not in `aspects`, while it is known that all
+current and future devices must support the aspects in `aspects`.  Lastly, the
+default JIT SPIR-V target (`spir64`) should not have an entry in the
+configuration file as it cannot guarantee anything about the devices supporting
+the target.
+
+When compiling a SYCL program, where $[t1, t2, \ldots, tn]$ are the $n$ targets
+specified in `-fsycl-targets` including any targets implicitly added by the
+driver, the driver defines the following macros:
+* `__SYCL_ALL_DEVICES_HAVE_`$i$`__` as `1` for all $i$ in
+${\bigcap}^n_{k=1} A^{all}_{tk}$.
+* `__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` as `1` if
+${\bigcup}^n_{k=1} A^{any}_{tk}$ is the set of all aspects.
+* `__SYCL_ANY_DEVICE_HAS_`$j$`__` as `1` for all $j$ in
+${\bigcup}^n_{k=1} A^{any}_{tk}$ if `__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` was not
+defined.
```

Comment:
It's not clear to me - would these macros be set for device compilation only?  Given usage, it does seem to be the case, but maybe call it out for clarity.

Reply:
They should be set for both host and device compilations.

Reply:
I tried clarifying this in 8f15db067f3fb1bc096f21199f17bb158e3a4cc5.